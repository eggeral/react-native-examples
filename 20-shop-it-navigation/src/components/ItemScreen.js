import React, { Component } from 'react';
import { Button, StyleSheet, View, Text, TextInput } from 'react-native';


// We do not keep the item as state here because it is already part of the state ot the ItemsScreen
export default class ItemScreen extends Component {

  static navigationOptions = ({ navigation }) => {
    return {
      title: 'Item',
      headerRight: () => {
        const item = navigation.getParam('item');
        return <Button
          disabled={item.name.length <= 0}
          onPress={() => {
            navigation.getParam('onOk')(item);
            navigation.goBack();
          }}
          title="Ok"
        />
      },
    }
  };

  state = { item: {}}

  componentDidMount() {
    const item = this.props.navigation.getParam('item');
    this.setState({ item: item })
  }

  changeItemName(name) {
    this.setState(state => {
      const newItem = { ...state.item, name: name };
      this.props.navigation.setParams({ item: newItem });
      return { item: newItem }
    });
  }

  changeItemCount(count) {
    this.setState(state => {
      const newItem = { ...state.item, count: count };
      this.props.navigation.setParams({ item: newItem });
      return { item: newItem }
    });
  }

  render() {
    return <View style={styles.container}>
      <TextInput
        style={styles.itemNameInput}
        value={this.state.item.name}
        placeholder="Add item"
        onChangeText={newText => this.changeItemName(newText)}
      />
      <View style={styles.countInputContainer}>
        <Button title="less"
          disabled={this.state.item.count <= 1}
          onPress={() => this.changeItemCount(this.state.item.count - 1)}
        ></Button>
        <Text style={styles.countText}>{this.state.item.count}</Text>
        <Button title="more"
          onPress={() => this.changeItemCount(this.state.item.count + 1)}
        ></Button>
      </View>
    </View>
  }
}

const styles = StyleSheet.create({

  container: {
    marginLeft: 10,
    marginRight: 10,
    marginTop: 10,
    marginBottom: 50,
    flex: 1,
    flexDirection: 'column',
  },

  itemNameInput: {
    fontSize: 16
  },

  countInputContainer: {
    marginTop: 25,
    flex: 0,
    flexDirection: 'row',
    alignItems: 'center'
  },

  countText: {
    fontSize: 16
  }

});