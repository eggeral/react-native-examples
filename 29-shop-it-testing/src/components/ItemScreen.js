import React, { Component } from 'react';
import { Button, StyleSheet, View, Text, TextInput, Image } from 'react-native';
import { QrCodeButton, CameraButton } from './Buttons';
export default class ItemScreen extends Component {

  static navigationOptions = ({ navigation }) => {
    return {
      title: 'Item',
      headerRight: () => {
        const item = navigation.getParam('item');
        return <Button
          disabled={item.name.length <= 0}
          onPress={() => {
            navigation.getParam('onOk')(navigation.getParam('item'));
            navigation.goBack();
          }}
          accessibilityLabel="okButton"
          title="Ok"
        />
      },
    }
  };

  state = { item: {} }

  componentDidMount() {
    const item = this.props.navigation.getParam('item');
    this.setState({ item: item })
  }

  changeItemName(name) {
    this.setState(state => {
      const newItem = { ...state.item, name: name };
      this.props.navigation.setParams({ item: newItem });
      return { item: newItem }
    });
  }

  changeItemCount(count) {
    this.setState(state => {
      const newItem = { ...state.item, count: count };
      this.props.navigation.setParams({ item: newItem });
      return { item: newItem }
    });
  }

  changeItemPhoto(photo) {
    this.setState(state => {
      const newItem = { ...state.item, photo: photo };
      this.props.navigation.setParams({ item: newItem });
      return { item: newItem }
    });
  }

  render() {
    return <View style={styles.container}>
      <View style={styles.nameInputContainer}>
        <TextInput
          style={styles.itemNameInput}
          value={this.state.item.name}
          placeholder="Add item"
          accessibilityLabel="itemNameInput"
          onChangeText={newText => this.changeItemName(newText)}
        />
        <QrCodeButton
          onPress={() => {
            this.props.navigation.navigate('QrCodeScanner', {
              onScanned: data => this.changeItemName(data)
            })
          }} />
      </View>
      <View style={styles.countInputContainer}>
        <Button title="less"
          disabled={this.state.item.count <= 1}
          accessibilityLabel="lessButton"
          onPress={() => this.changeItemCount(this.state.item.count - 1)}
        ></Button>
        <Text style={styles.countText}>{this.state.item.count}</Text>
        <Button title="more"
          accessibilityLabel="moreButton"
          onPress={() => this.changeItemCount(this.state.item.count + 1)}
        ></Button>
      </View>
      <View style={styles.takePhotoContainer}>
        <CameraButton
          accessibilityLabel="takePhotoButton"
          onPress={() => this.props.navigation.navigate('TakePhoto', {
            onTaken: data => this.changeItemPhoto(data)
          })} />
        <Image style={styles.photo} source={{ uri: this.state.item.photo }}></Image>
        {this.state.item.photo &&
          <Button title="X"
            accessibilityLabel="clearPhotoButton"
            onPress={() => this.changeItemPhoto(undefined)}
          ></Button>
        }
      </View>
    </View>
  }
}

const styles = StyleSheet.create({

  container: {
    marginLeft: 10,
    marginRight: 10,
    marginTop: 10,
    marginBottom: 50,
    flex: 1,
    flexDirection: 'column',
  },

  nameInputContainer: {
    marginTop: 0,
    flex: 0,
    flexDirection: 'row',
  },

  itemNameInput: {
    fontSize: 16,
    flex: 1
  },

  countInputContainer: {
    marginTop: 10,
    flex: 0,
    flexDirection: 'row',
    alignItems: 'center'
  },

  takePhotoContainer: {
    marginTop: 10,
    flex: 1,
    alignItems: 'center'
  },

  countText: {
    fontSize: 16
  },

  photo: {
    flex: 1,
    marginTop: 10,
    width: 300,
    height: 100,
    resizeMode: 'contain',
  }

});