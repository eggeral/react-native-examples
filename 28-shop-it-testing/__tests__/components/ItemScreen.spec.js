jest.mock('../../src/components/Buttons');
import React from 'react';
import { cleanup, fireEvent, render } from '@testing-library/react-native';
import renderer from 'react-test-renderer';
import ItemScreen from '../../src/components/ItemScreen';

describe('ItemScreen', () => {

    afterEach(cleanup);

    it('should render correctly', () => {
        const item = { id: 1, name: 'item 1' };
        const navigation = {
            getParam: () => item
        }
        const tree = renderer.create(<ItemScreen navigation={navigation} />).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('can change the item', () => {
        // given
        const item = { id: 1, name: 'item 1', count: 1, photo: 'base64:photo' };
        let changedItem
        let navigationParams = {
            item,
            onOk: item => changedItem = item
        }
        const navigation = {
            getParam: paramName => navigationParams[paramName],
            setParams: it => Object.assign(navigationParams, it),
            goBack: () => { }
        }
        const component = render(<ItemScreen navigation={navigation} />);
        const header = render(ItemScreen.navigationOptions({ navigation }).headerRight());
        const itemNameInput = component.getByLabelText('itemNameInput');
        const lessButton = component.getByLabelText('lessButton');
        const moreButton = component.getByLabelText('moreButton');
        const clearPhotoButton = component.getByLabelText('clearPhotoButton');
        const okButton = header.getByLabelText('okButton');

        // when
        fireEvent.changeText(itemNameInput, "test");
        fireEvent.press(moreButton);
        fireEvent.press(moreButton);
        fireEvent.press(lessButton);
        fireEvent.press(clearPhotoButton);

        fireEvent.press(okButton);

        // then
        expect(navigationParams.item).toEqual({ id: 1, name: 'test', count: 2 });
        expect(changedItem).toEqual({ id: 1, name: 'test', count: 2 });

    });

    it('can scan the item name from a QR code', () => {
        // given
        const item = { id: 1, name: 'item 1' };
        let navigationParams = {
            item,
            onOk: item => changedItem = item
        };
        let currentScreen;
        const navigation = {
            getParam: paramName => navigationParams[paramName],
            setParams: it => Object.assign(navigationParams, it),
            navigate: (title, params) => currentScreen = { title, params },
            goBack: () => { }
        }
        const component = render(<ItemScreen navigation={navigation} />);
        const qrCodeButton = component.getByLabelText('qrCodeButton');

        // when
        fireEvent.press(qrCodeButton);
        expect(currentScreen.title).toEqual('QrCodeScanner');
        currentScreen.params.onScanned('qr code');

        // then
        expect(navigationParams.item).toEqual({ id: 1, name: 'qr code' });

    });

    it('can add a photo to  an item', () => {
        // given
        const item = { id: 1, name: 'item 1' };
        let navigationParams = {
            item,
            onOk: item => changedItem = item
        };
        let currentScreen;
        const navigation = {
            getParam: paramName => navigationParams[paramName],
            setParams: it => Object.assign(navigationParams, it),
            navigate: (title, params) => currentScreen = { title, params },
            goBack: () => { }
        }
        const component = render(<ItemScreen navigation={navigation} />);
        const cameraButton = component.getByLabelText('cameraButton');

        // when
        fireEvent.press(cameraButton);
        expect(currentScreen.title).toEqual('TakePhoto');
        currentScreen.params.onTaken('image of a tree');

        // then
        expect(navigationParams.item).toEqual({ id: 1, name: 'item 1', photo: 'image of a tree' });

    });

})