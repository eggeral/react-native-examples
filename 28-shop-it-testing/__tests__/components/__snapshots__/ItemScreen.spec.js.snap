// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ItemScreen should render correctly 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "flexDirection": "column",
      "marginBottom": 50,
      "marginLeft": 10,
      "marginRight": 10,
      "marginTop": 10,
    }
  }
>
  <View
    style={
      Object {
        "flex": 0,
        "flexDirection": "row",
        "marginTop": 0,
      }
    }
  >
    <TextInput
      accessibilityLabel="itemNameInput"
      allowFontScaling={true}
      onChangeText={[Function]}
      placeholder="Add item"
      rejectResponderTermination={true}
      style={
        Object {
          "flex": 1,
          "fontSize": 16,
        }
      }
      underlineColorAndroid="transparent"
      value="item 1"
    />
    <View
      accessibilityLabel="qrCodeButton"
      accessibilityRole="button"
      accessibilityStates={Array []}
      accessible={true}
      focusable={true}
      isTVSelectable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Array [
            Object {},
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#007AFF",
                "fontSize": 18,
                "padding": 8,
                "textAlign": "center",
              },
            ]
          }
        >
          QrCodeButton
        </Text>
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 0,
        "flexDirection": "row",
        "marginTop": 10,
      }
    }
  >
    <View
      accessibilityLabel="lessButton"
      accessibilityRole="button"
      accessibilityStates={Array []}
      accessible={true}
      focusable={true}
      isTVSelectable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Array [
            Object {},
          ]
        }
      >
        <Text
          disabled={false}
          style={
            Array [
              Object {
                "color": "#007AFF",
                "fontSize": 18,
                "padding": 8,
                "textAlign": "center",
              },
            ]
          }
        >
          less
        </Text>
      </View>
    </View>
    <Text
      style={
        Object {
          "fontSize": 16,
        }
      }
    />
    <View
      accessibilityLabel="moreButton"
      accessibilityRole="button"
      accessibilityStates={Array []}
      accessible={true}
      focusable={true}
      isTVSelectable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Array [
            Object {},
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#007AFF",
                "fontSize": 18,
                "padding": 8,
                "textAlign": "center",
              },
            ]
          }
        >
          more
        </Text>
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "marginTop": 10,
      }
    }
  >
    <View
      accessibilityLabel="cameraButton"
      accessibilityRole="button"
      accessibilityStates={Array []}
      accessible={true}
      focusable={true}
      isTVSelectable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Array [
            Object {},
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#007AFF",
                "fontSize": 18,
                "padding": 8,
                "textAlign": "center",
              },
            ]
          }
        >
          CameraButton
        </Text>
      </View>
    </View>
    <Image
      source={
        Object {
          "uri": undefined,
        }
      }
      style={
        Object {
          "flex": 1,
          "height": 100,
          "marginTop": 10,
          "resizeMode": "contain",
          "width": 300,
        }
      }
    />
  </View>
</View>
`;
